test/unix/test_sleep_and_timeout.ml

```diff
@@ -0,0 +1,80 @@
+(* This file is part of Lwt, released under the MIT license. See LICENSE.md for
+   details, or visit https://github.com/ocsigen/lwt/blob/master/LICENSE.md. *)
+
+open Test
+open Lwt.Infix
+
+let cmp_elapsed_time start_time expected_time =
+  let elapsed_time = Unix.gettimeofday () -. start_time in
+  elapsed_time -. expected_time < 0.005
+
+let suite = suite "Lwt_unix sleep and timeout" [
+    test "sleep" begin fun () ->
+      let start_time = Unix.gettimeofday () in
+      let duration = 1.0 in
+      Lwt_unix.sleep duration
+      >>= fun () ->
+      let check = cmp_elapsed_time start_time duration in
+      Lwt.return check
+    end;
+
+    test "auto_yield" begin fun () ->
+      let start_time = Unix.gettimeofday () in
+      let duration = 1.0 in
+      let wait = Lwt_unix.auto_yield duration in
+      wait ()
+      >>= fun () ->
+      let check = cmp_elapsed_time start_time duration in
+      Lwt.return check
+    end;
```

Comment:
`let f = auto_yield timeout in` is something like:

- If it has been less than `timeout` time since the last time `f ()` was called, behave like `Lwt.return ()`.
- If it has been more than `timeout` time since the last time `f ()` was called, behave like `Lwt.yield ()`.

You can test whether `f ()` is yielding or not by comparing elapsed times. I think you want something like call `f ()`, it shouldn't yield (time should be near 0). Wait one second, call `f ()` again. It should now yield (time near 1). Call `f ()` immediately again, it should not yield. I'd then call it immediately a second time, and then wait one second and check for a second yield, just to make sure this is truly repeating.

Also, the docs of this function need improving. I'll open a separate issue.

This test I think doesn't "work" at all. I suspect the elapsed time is close to zero, and the test is succeeding because `cmp_elapsed_time` allows the elapsed time to be less than the duration right now.

Reply:
Opened #631.

Reply:
> This test I think doesn't "work" at all. I suspect the elapsed time is close to zero, and the test is succeeding because cmp_elapsed_time allows the elapsed time to be less than the duration right now

yes, that right.